pkg/ring/kv/consul/client.go

```diff
@@ -264,3 +289,32 @@ func (c *Client) Get(ctx context.Context, key string) (interface{}, error) {
 	}
 	return c.codec.Decode(kvp.Value)
 }
+
+func checkLastIndex(index, metaLastIndex uint64) (newIndex uint64, skip bool) {
+	// See https://www.consul.io/api/features/blocking.html#implementation-details for logic behind these checks.
+	if metaLastIndex == 0 {
+		// Don't just keep using index=0.
+		// After blocking request, returned index must be at least 1.
+		return 1, false
+	} else if metaLastIndex < index {
+		// Index reset.
+		return 0, false
+	} else if index == metaLastIndex {
+		// Skip if the index is the same as last time, because the key value is
+		// guaranteed to be the same as last time
+		return metaLastIndex, true
+	} else {
+		return metaLastIndex, false
+	}
+}
+
+func (c *Client) createRateLimiter() *rate.Limiter {
+	if c.cfg.WatchKeyRate <= 0 {
+		return nil
+	}
+	burst := c.cfg.WatchKeyBurst
+	if burst < 1 {
```

Comment:
Oh, I was about to add this check. I think it makes sense, but you seem to have changed your mind?

Reply:
I've rechecked the rate limiter implementation and it's not an issue as far as the burst is at least 1, because here you always consume 1 token each time.

Reply:
What I mean is that we always use `limiter.Wait()` here, which internally calls `WaitN()` with `n = 1` (`n` is the number of tokens to reserve), while `burst` is the max number of tokens you can consume in a single operation, which in our case is always `1`, so an hard-coded `1` value should be fine (if I'm not missing anything).

Reply:
Burst being the max number of tokens you can consume is not the only reason to use burst. Another is to temporarily raise the limit, if limiter wasn't used recently (or at the beginning of the operation). See here: https://play.golang.org/p/zBtoOUf5fSl ... notice how first 5 operations are allowed immediately, and only subsequent operations are limited to 1 per second.

Reply:
You're right.

Reply:
On the other hand, you're right in that checking that burst is higher than limit doesn't help much – if you use limit 5, and burst 1, client using Wait() is still allowed 5 operations / second. So all is fine here. I will resolve it now.